packages/website/blog/2024-03-25-changes-to-consistent-type-imports-with-decorators.md

```diff
@@ -0,0 +1,133 @@
+---
+authors:
+  - image_url: /img/team/bradzacher.jpg
+    name: Brad Zacher
+    title: typescript-eslint Maintainer
+    url: https://github.com/bradzacher
+description: Changes to consistent-type-imports when used with decorators, experimentalDecorators, and emitDecoratorMetadata
+slug: changes-to-consistent-type-imports-with-decorators
+tags:
+  [
+    consistent-type-imports,
+    experimentalDecorators,
+    emitDecoratorMetadata,
+    typescript-eslint,
+  ]
+title: Changes to `consistent-type-imports` with Legacy Decorators and Decorator Metadata
+---
+
+We've made some changes to the [`consistent-type-imports` rule](/rules/consistent-type-imports) to fix some long-standing issues when used alongside `experimentalDecorators: true` and `emitDecoratorMetadata: true`. These changes increase safety and prevent invalid fixes when using decorator metadata.
+
+<!-- truncate -->
+
+## Experimental Decorator Metadata
+
+TypeScript's [`experimentalDecorators` compiler option](https://aka.ms/tsconfig#experimentalDecorators) (referred to as "legacy decorators" from here on) turns on support for an old version of the [JavaScript TC39 decorator proposal](https://github.com/tc39/proposal-decorators) that was never standardized. TypeScript's legacy decorators are similar to the current proposal, but differ in that they use [metadata reflection](https://rbuckton.github.io/reflect-metadata/) when TypeScript's [`emitDecoratorMetadata` compiler option](https://aka.ms/tsconfig#emitDecoratorMetadata) is turned on.
+
+When using legacy decorators with decorator metadata and a class is annotated with decorators, TypeScript will emit runtime metadata for the class (see the example below). That decorator metadata will capture property types, method parameter types, and method return types. Decorator metadata provides a bridge between the types (which are not available at compile time) and the runtime code.
+
+The downside of generating this runtime code that it is derived using type information: meaning that the runtime code emitted changes based on the cross-file type information that TypeScript has computed. Doing so violates a key [TypeScript design goal](https://github.com/microsoft/TypeScript/wiki/TypeScript-Design-Goals) of not changing runtime behavior based on type information.
+
+To illustrate what this means consider the following snippet:
+
+```ts
+import Foo from 'foo';
+import decorator from 'decorator';
+
+class Clazz {
+  @decorator
+  method(arg: Foo) {}
+}
+```
+
+TypeScript will transpile this code to the following:
+
+<!-- prettier-ignore -->
+```ts
+var _a;
```

Comment:
Nit: this isn't used later.

```suggestion
```

If TypeScript is producing an `_a` despite never using it, then that's a bug we can ignore.

Reply:
Oh nah it was there cos the code has no types so this is an extra case I (purposely) didn't mention. In this case when TS can't determine the type but it can see its a value-import - instead TS emits  `[typeof _a = (typeof Foo !== "undefined" && Foo) === "function" ? _a : Object]`. Which is some funky code-golf heh.

Can remove it though yeah as it's not necessary for the examples I'm mentioning.